jmdaemon/jmdaemon/onionmc.py

```diff
@@ -0,0 +1,1214 @@
+        encoded_msg = OnionCustomMessage(privmsg,
+                        JM_MESSAGE_TYPES["privmsg"])
+        self._send(peer, encoded_msg)
+        # If possible, we forward the from-nick's network location
+        # to the to-nick peer, so they can just talk directly next time.
+        peer_from = self.get_peer_by_nick(from_nick)
+        if not peer_from:
+            return
+        self.send_peers(peer, peer_filter=[peer_from])
+
+    def process_control_message(self, peerid: str, msgtype: int,
+                                msgval: str) -> bool:
+        """ Triggered by a directory node feeding us
+        peers, or by a connect/disconnect hook; this is our housekeeping
+        to try to create, and keep track of, useful connections.
+        The returned boolean indicates whether we succeeded in processing
+        the message or whether it must be analyzed again (note e.g. that
+        we return True for a rejected message!)
+        """
+        all_ctrl = list(LOCAL_CONTROL_MESSAGE_TYPES.values(
+            )) + list(CONTROL_MESSAGE_TYPES.values())
+        if msgtype not in all_ctrl:
+            return False
+        # this is too noisy, but TODO, investigate allowing
+        # some kind of control message monitoring e.g. default-off
+        # log-to-file (we don't currently have a 'TRACE' level debug).
+        #log.debug("received control message: {},{}".format(msgtype, msgval))
+        if msgtype == CONTROL_MESSAGE_TYPES["peerlist"]:
+            # This is the base method of seeding connections;
+            # a directory node can send this any time.
+            # These messages can only be accepted from directory peers
+            # (which we have configured ourselves):
+            peer = self.get_peer_by_id(peerid)
+            if not peer or not peer.directory:
+                return True
+            try:
+                peerlist = msgval.split(",")
+                for peer in peerlist:
+                    # defaults mean we just add the peer, not
+                    # add or alter its connection status:
+                    self.add_peer(peer, with_nick=True)
+            except Exception as e:
+                log.debug("Incorrectly formatted peer list: {}, "
+                      "ignoring, {}".format(msgval, e))
+            # returning True whether raised or not - see docstring
+            return True
+        elif msgtype == CONTROL_MESSAGE_TYPES["getpeerlist"]:
+            # getpeerlist must be accompanied by a full node
+            # locator, and nick;
+            # add that peer before returning our peer list.
+            # First, this should only be received by a directory
+            if not self.self_as_peer.directory:
+                return True
+            p = self.add_peer(msgval, connection=True,
+                              overwrite_connection=True, with_nick=True)
+            try:
+                self.send_peers(p)

```

Comment:
Yes good catch this is a result of us not currently using `getpeerlist`. I'm tempted to just disable it for now since we don't want clients just willy-nilly requesting *all* peers when the existing "send peer if they privmsged" is probably good enough.

Reply:
`getpeerlist` just ignored for now as of https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1182/commits/2966985b9104905b9fad263342e21a608a4fdf5e

(can always be reintroduced later, but if we do it'll be a new protocol version, and one in which the implication of an arbitrarily large peer list request is considered).